introspection-engine/connectors/sql-introspection-connector/src/introspection_helpers.rs

```diff
@@ -302,58 +303,35 @@ pub(crate) fn calculate_backrelation_field(
     }
 }
 
-pub(crate) fn is_id(column: &sql::Column, table: &sql::Table) -> bool {
-    table
-        .primary_key
-        .as_ref()
-        .map(|pk| pk.is_single_primary_key(&column.name))
-        .unwrap_or(false)
-}
-
-pub(crate) fn calculate_relation_name(
-    schema: &SqlSchema,
-    fk: &ForeignKey,
-    table: &Table,
+// This is not called for prisma many to many relations. For them the name is just the name of the join table.
+fn calculate_relation_name(
+    fk: ForeignKeyWalker<'_>,
     m2m_table_names: &[String],
-) -> Result<String, SqlError> {
-    // This is not called for prisma many to many relations. For them the name is just the name of the join table.
-    let referenced_model = &fk.referenced_table;
-    let model_with_fk = &table.name;
-    let fk_column_name = fk.columns.join("_");
-
-    let mut fk_to_same_model: Vec<ForeignKey> = table
-        .foreign_keys
-        .clone()
-        .into_iter()
-        .filter(|fk| &fk.referenced_table == referenced_model)
-        .collect();
-
-    fk_to_same_model.clear_duplicates();
-
-    match schema.table(referenced_model) {
-        Err(table_name) => Err(SqlError::SchemaInconsistent {
-            explanation: format!("Table {} not found.", table_name),
-        }),
-        Ok(other_table) => {
-            let fk_from_other_model_to_this_exist = other_table
-                .foreign_keys
-                .iter()
-                .any(|fk| &fk.referenced_table == model_with_fk);
-
-            let unambiguous_name = RelationNames::name_for_unambiguous_relation(model_with_fk, referenced_model);
-
-            // this needs to know whether there are m2m relations and then use ambiguous name path
-            let name = if fk_to_same_model.len() < 2
-                && !fk_from_other_model_to_this_exist
-                && !m2m_table_names.contains(&unambiguous_name)
-            {
-                unambiguous_name
-            } else {
-                RelationNames::name_for_ambiguous_relation(model_with_fk, referenced_model, &fk_column_name)
-            };
-
-            Ok(name)
-        }
+    duplicated_foreign_keys: &HashSet<sql::ForeignKeyId>,
+) -> String {
+    let referenced_model = fk.referenced_table().name();
+    let model_with_fk = fk.table().name();
+    let fk_column_name = fk.constrained_column_names().join("_");
+    let name_is_ambiguous = fk.schema.walk_foreign_keys().any(|other_fk| {
+        let both_ids = [fk.referenced_table().id, fk.table().id];
+
+        other_fk.id != fk.id
+            && !duplicated_foreign_keys.contains(&other_fk.id)
+            && (
+                // Self-relation case
+                (other_fk.referenced_table().id == other_fk.table().id && other_fk.table().id == fk.table().id)
```

Comment:
What if fk references another table?

Reply:
We don't know that. That seems wrong. Referenced and current tables are all the same => then it's a self relation.